plugins/modem-manager/fu-dfota-updater.c

```diff
@@ -0,0 +1,382 @@
+				    "could not match size and checksum");
+		return FALSE;
+	}
+
+	size_match = g_match_info_fetch(match_info, 1);
+	checksum_match = g_match_info_fetch(match_info, 2);
+
+	g_debug("parsed checksum '%s' and size '%s'", checksum_match, size_match);
+
+	*checksum = g_steal_pointer(&checksum_match);
+
+	if (!g_ascii_string_to_unsigned(size_match, 10, 0, G_MAXSIZE, size, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+gboolean
+fu_dfota_updater_upload_firmware(FuDfotaUpdater *self, GBytes *fw, GError **error)
+{
+	g_autoptr(FuChunkArray) chunks = fu_chunk_array_new_from_bytes(fw, 0x0, 0x800);
+	guint chunk_count = fu_chunk_array_length(chunks);
+	g_autofree gchar *checksum = fu_dfota_updater_compute_checksum(fw);
+	g_autofree gchar *checksum_parsed = NULL;
+	gsize size = g_bytes_get_size(fw);
+	gsize size_parsed = 0;
+
+	g_debug("expecting checksum '%s' and size '%" G_GSIZE_FORMAT "'", checksum, size);
+
+	for (guint i = 0; i < chunk_count; i++) {
+		g_autoptr(FuChunk) chk = fu_chunk_array_index(chunks, i, error);
+
+		if (chk == NULL)
+			return FALSE;
+
+		if (!fu_dfota_updater_upload_chunk(self, chk, error)) {
+			g_prefix_error(error, "failed at chunk %u: ", i);
+			return FALSE;
+		}
+		if (i % 100 == 0)
+			g_debug("wrote chunk %u/%u", i, chunk_count - 1);
+	}
+
+	if (!fu_dfota_updater_parse_upload_result(self, &checksum_parsed, &size_parsed, error))
+		return FALSE;
+
+	if (size != size_parsed) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "firmware size mismatch - expected %" G_GSIZE_FORMAT
+			    ", but was %" G_GSIZE_FORMAT,
+			    size,
+			    size_parsed);
+		return FALSE;
+	}
+	if (g_strcmp0(checksum, checksum_parsed) != 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "checksum mismatch - expected %s, but was %s",
+			    checksum,
+			    checksum_parsed);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+gboolean
+fu_dfota_updater_open(FuDfotaUpdater *self, GError **error)
+{
+	if (self->io_channel == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "no channel provided for update");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+gboolean
+fu_dfota_updater_close(FuDfotaUpdater *self, GError **error)
+{
+	/* closing will be handled by locker */
+	return TRUE;
+}
+
+static gboolean
+fu_dfota_updater_parse_fota_response(gchar *response,
+				     FuProgress *progress,
+				     gboolean *finished,
+				     GError **error)
+{
+	g_autoptr(GRegex) fota_regex = NULL;
+	g_autoptr(GMatchInfo) match_info = NULL;
+	g_autofree gchar *status_match = NULL;
+	g_autofree gchar *status_number_match = NULL;
+	guint64 status_number = 0;
+
+	/* +QIND: "FOTA","<STATUS>"(,<number>)? */
+	fota_regex = g_regex_new("\\+QIND:\\s*\"FOTA\",\"([A-Z]+)\"(,(\\d+))?", 0, 0, NULL);
+
+	if (!g_regex_match(fota_regex, response, 0, &match_info)) {
+		g_warning("got unexpected response '%s'", response);
```

Comment:
shouldn't this be more fatal?

Reply:
The device seems to send some send some arbitrary message between status outputs, so we don't want to abort directly if the regex does not match. I'm still investigating if this is really an issue with the device

Reply:
The device itself reproducibly sends 1-2 incomplete messages for an update, so we have to continue the update if there are unexpected messages. I've also added a comment in the source code to clarify.

Reply:
From a user perspective a `g_warning` will show up in the journal and make people think something is wrong.  I think it should at least downgrade to `g_info` or `g_debug` then.